### ISSUE TITLE
19b: LLM inference script (method4_generate.py) for cluster GPU

### ISSUE THREAD

**Issue by ofekvs18** [ofekvs18/UnderGradProject#21]
**Parent**: #19
**Depends on**: #20 (prompts must be designed first)

### Goal

Create `src/method4_generate.py` — the GPU inference script that loads Med-Gemma 4B, runs all prompt configs, and saves raw LLM outputs to JSON. This script runs on the university cluster GPU node.

### Implementation

```python
"""
Method 4: LLM Formula Generation — Inference Script
Runs on GPU. Loads Med-Gemma 4B, generates formulas, saves raw outputs.

Usage:
    python src/method4_generate.py                    # run all configs
    python src/method4_generate.py --config blind_focused  # run one config
    python src/method4_generate.py --dry-run           # print prompts, don't run inference
"""

import json
import time
import argparse
from pathlib import Path
from method4_prompts import get_all_prompt_configs

RESULTS_DIR = Path("results/method4_llm")
MODEL_NAME = "google/medgemma-4b-it"  # instruction-tuned variant

def load_model():
    """Load Med-Gemma 4B with HuggingFace transformers."""
    from transformers import AutoTokenizer, AutoModelForCausalLM
    import torch

    tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
    model = AutoModelForCausalLM.from_pretrained(
        MODEL_NAME,
        torch_dtype=torch.bfloat16,  # or float16, depending on GPU
        device_map="auto",
    )
    return model, tokenizer

def generate_formulas(model, tokenizer, prompt, temperature=0.5, max_new_tokens=2048):
    """Run one inference call. Returns raw text output."""
    ...

def run_all_configs(model, tokenizer, configs, n_repeats=4):
    """Run all prompt configs with repeats. Returns list of result dicts."""
    results = []
    for name, prompt, temp in configs:
        for repeat_idx in range(n_repeats):
            print(f"  Running {name} (repeat {repeat_idx+1}/{n_repeats}, temp={temp})...")
            start = time.time()
            raw_output = generate_formulas(model, tokenizer, prompt, temperature=temp)
            elapsed = time.time() - start
            results.append({
                "config_name": name,
                "repeat_idx": repeat_idx,
                "temperature": temp,
                "prompt": prompt,
                "raw_output": raw_output,
                "elapsed_seconds": elapsed,
                "model": MODEL_NAME,
            })
            print(f"    Done in {elapsed:.1f}s, output length: {len(raw_output)} chars")
    return results

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--config", type=str, default=None, help="Run only this config name")
    parser.add_argument("--dry-run", action="store_true", help="Print prompts without running inference")
    parser.add_argument("--repeats", type=int, default=4, help="Number of repeats per config")
    args = parser.parse_args()

    configs = get_all_prompt_configs()
    if args.config:
        configs = [(n, p, t) for n, p, t in configs if n == args.config]
        assert configs, f"Config '{args.config}' not found"

    if args.dry_run:
        for name, prompt, temp in configs:
            print(f"\n{'='*60}\nConfig: {name} (temp={temp})\n{'='*60}")
            print(prompt)
        return

    print(f"Loading {MODEL_NAME}...")
    model, tokenizer = load_model()
    print("Model loaded.")

    RESULTS_DIR.mkdir(parents=True, exist_ok=True)
    results = run_all_configs(model, tokenizer, configs, n_repeats=args.repeats)

    output_path = RESULTS_DIR / "raw_outputs.json"
    with open(output_path, "w") as f:
        json.dump(results, f, indent=2)
    print(f"\nSaved {len(results)} outputs to {output_path}")

if __name__ == "__main__":
    main()
```

### Key design decisions

**Model loading**: Use `device_map="auto"` for automatic GPU placement. Med-Gemma 4B should fit on a single GPU (~8-10GB VRAM in bfloat16).

**Repeats**: Each config runs 4 times by default (with temperature > 0, each run produces different formulas). This gives diversity without excessive compute.

**Saving strategy**: Save ALL raw outputs including the full prompt and timing. This makes debugging easier — you can see exactly what the model received and produced.

**Dry-run mode**: `--dry-run` prints prompts without loading the model. Essential for verifying prompts on a CPU-only machine before submitting the GPU job.

**Single config mode**: `--config blind_focused` runs just one config. Useful for testing or re-running a failed config.

### Cluster job script

The issue should also create a SLURM/PBS job script template (adjust for your cluster's scheduler):

```bash
#!/bin/bash
#SBATCH --job-name=method4_llm
#SBATCH --partition=gpu
#SBATCH --gres=gpu:1
#SBATCH --mem=24G
#SBATCH --time=02:00:00
#SBATCH --output=logs/method4_%j.out

module load python/3.10
module load cuda/12.1

cd ~/biomarker-pipeline
source venv/bin/activate  # or conda activate

python src/method4_generate.py --repeats 4
```

### Output

```
results/method4_llm/
└── raw_outputs.json    # list of dicts, one per inference call
```

Each entry in `raw_outputs.json`:
```json
{
    "config_name": "blind_focused",
    "repeat_idx": 0,
    "temperature": 0.3,
    "prompt": "You are a medical researcher...",
    "raw_output": "np.log1p(rdw) * mchc / (plt + 1)\n(wbc * rdw) / hgb\n...",
    "elapsed_seconds": 45.2,
    "model": "google/medgemma-4b-it"
}
```

### Dependencies

Add to `requirements.txt`:
```
transformers>=4.40.0
torch>=2.1.0
accelerate
```

**Note**: These are GPU-only dependencies. The evaluation script (#22) doesn't need them.

### Verification

1. `--dry-run` mode works and prints all 6 prompt configs
2. Running with `--config blind_focused --repeats 1` produces a valid `raw_outputs.json` with one entry
3. The raw_output field contains text that looks like formula candidates (even if parsing is messy)
4. Model loads successfully on the cluster GPU node
5. Total runtime for all 24 calls should be under 2 hours

### What NOT to do

- Do NOT parse or evaluate formulas in this script (that's #22)
- Do NOT use the test set or any data — this script only does LLM inference
- Do NOT hardcode prompts — import from `method4_prompts.py`
- Do NOT try to filter or clean LLM output — save everything raw
